### PR TITLE
Basic loading to/reloading from disk functionality

### PR DESCRIPTION
**Pull request by justinyum**
Exactly what it says on the tin. Two new functions "persist_object" and "get_persisted_object", along with helper function "object_id_to_persist_path". Tested these functions in the basic-persistence branch, wherein these functions were functional (issues were caused by the eviction policy). 

The hope is to integrate these 2 functions into the eviction policy once we better define the semantics of the LRU cache and eviction.

### CODE REVIEW

File: src/plasma/plasma_client.c [hunk 1]
```diff
@@ -539,3 +539,19 @@ int plasma_wait(plasma_connection *conn,
 int get_manager_fd(plasma_connection *conn) {
   return conn->manager_conn;
 }
+
+void plasma_persist(plasma_connection *conn, object_id object_id) {
+  plasma_request req = make_plasma_request(object_id);
+  plasma_send_request(conn->store_conn, PLASMA_PERSIST, &req);
+  if (conn->manager_conn >= 0) {
```
Comment (pcmoritz) [inline src/plasma/plasma_client.c]: Do we need to inform the manager here?

File: src/plasma/plasma_client.c [hunk 2]
```diff
@@ -539,3 +539,19 @@ int plasma_wait(plasma_connection *conn,
...
+    plasma_send_request(conn->manager_conn, PLASMA_PERSIST, &req);
+  }
+}
+
+void plasma_reload(plasma_connection *conn, object_id object_id) {
+  plasma_request req = make_plasma_request(object_id);
+  plasma_send_request(conn->store_conn, PLASMA_RELOAD, &req);
```
Comment (pcmoritz) [inline src/plasma/plasma_client.c]: same here

File: src/plasma/plasma_store.c [hunk 3]
```diff
@@ -359,6 +363,59 @@ void remove_objects(plasma_store_state *plasma_state,
   }
 }
 
+/* Convert object_id to a unique file path on-disk */
+char *object_id_to_persist_path(object_id object_id) {
```
Comment (pcmoritz) [inline src/plasma/plasma_store.c]: please use the function sha1_to_hex http://github.com/ray-project/ray/blob/master/src/common/common.h#L118

File: src/plasma/plasma_store.c [hunk 4]
```diff
@@ -33,6 +35,8 @@
 #include "malloc.h"
 #include "plasma_store.h"
 
+const char *PERSIST_PATH = "/Users/belugajustin/Downloads/misaka/persisted/";
```
Comment (pcmoritz) [inline src/plasma/plasma_store.c]: make this configurable by the user by passing in an option into the plasma store

File: src/plasma/plasma_store.c [hunk 5]
```diff
@@ -359,6 +363,59 @@ void remove_objects(plasma_store_state *plasma_state,
...
+  plasma_store_state *plasma_state = client_context->plasma_state;
+  object_table_entry *entry;
+  HASH_FIND(handle, plasma_state->plasma_store_info->objects, &object_id,
+            sizeof(object_id), entry);
+  CHECKM(entry != NULL, "To persist an object it must exist.");
+  CHECKM(entry->state == SEALED, "To persist an object it must be sealed.");
+  /* Check if the file containing the object already exists. */
+  char *file_path = object_id_to_persist_path(object_id);
+  struct stat buffer;
+  //CHECKM(stat(file_path, &buffer) != 0, "Cannot persist an object twice");
```
Comment (pcmoritz) [inline src/plasma/plasma_store.c]: please get rid of stale comments

File: src/plasma/plasma_store.c [hunk 6]
```diff
@@ -359,6 +363,59 @@ void remove_objects(plasma_store_state *plasma_state,
...
+            sizeof(object_id), entry);
+  CHECKM(entry != NULL, "To persist an object it must exist.");
+  CHECKM(entry->state == SEALED, "To persist an object it must be sealed.");
+  /* Check if the file containing the object already exists. */
+  char *file_path = object_id_to_persist_path(object_id);
+  struct stat buffer;
+  //CHECKM(stat(file_path, &buffer) != 0, "Cannot persist an object twice"); 
+  //took out b/c exposing persistence to the user now
+  /* Create the file if it does not exist and write object. */
+  int fd = open(file_path, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
```
Comment (pcmoritz) [inline src/plasma/plasma_store.c]: write a short comment about the flags

File: src/plasma/plasma_store.c [hunk 7]
```diff
@@ -359,6 +363,59 @@ void remove_objects(plasma_store_state *plasma_state,
...
+  CHECKM(fd > 0, "Something went wrong while creating persistence file");
+  write(fd, entry->pointer, entry->info.data_size + entry->info.metadata_size);
+  close(fd);
+  free(file_path);
+}
+
+/* Read a persisted object from disk back into memory. */
+void get_persisted_object(client *client_context, object_id object_id) {
```
Comment (pcmoritz) [inline src/plasma/plasma_store.c]: rename this to load_object or something like this

File: src/plasma/plasma_store.c [hunk 8]
```diff
@@ -359,6 +363,59 @@ void remove_objects(plasma_store_state *plasma_state,
...
+  LOG_DEBUG("reading persisted object");
+  plasma_store_state *plasma_state = client_context->plasma_state;
+  object_table_entry *entry;
+  HASH_FIND(handle, plasma_state->plasma_store_info->objects, &object_id,
+            sizeof(object_id), entry);
+  CHECKM(entry != NULL, "There must be an object entry to read it into.");
```
Comment (pcmoritz) [inline src/plasma/plasma_store.c]: put an if here, and call plasma_create with the right size to create the object. Also make sure to write metadata size and object size to the file (can use a struct for this, plasma_persistence_data with a size and a metadata_size field or something like this).

File: src/plasma/plasma_store.h [hunk 9]
```diff
@@ -89,4 +89,7 @@ void remove_objects(plasma_store_state *plasma_state,
                     int64_t num_objects_to_evict,
                     object_id *objects_to_evict);
 
+void persist_object(client *client_context, object_id object_id);
+void get_persisted_object(client *client_context, object_id object_id);
```
Comment (pcmoritz) [inline src/plasma/plasma_store.h]: maybe name these persist_object and load_object?

File: src/plasma/lib/python/plasma.py [hunk 10]
```diff
@@ -195,6 +195,26 @@ def seal(self, object_id):
     """
     self.client.plasma_seal(self.plasma_conn, make_plasma_id(object_id))
 
+  def persist(self, object_id):
```
Comment (pcmoritz) [inline src/plasma/lib/python/plasma.py]: how about persist_object and load_object here?

File: src/plasma/test/test.py [hunk 11]
```diff
@@ -101,6 +102,24 @@ def test_create(self):
     for i in range(length):
       self.assertEqual(memory_buffer[i], chr(i % 256))
 
+  def test_persist_sanity(self):
```
Comment (pcmoritz) [inline src/plasma/test/test.py]: once you implemented the reload if the object doesn't exist already, please also write a test case for this.

File: src/plasma/plasma_store.h [hunk 12]
```diff
@@ -89,4 +89,7 @@ void remove_objects(plasma_store_state *plasma_state,
                     int64_t num_objects_to_evict,
                     object_id *objects_to_evict);
 
+void persist_object(client *client_context, object_id object_id);
```
Comment (robertnishihara) [inline src/plasma/plasma_store.h]: document these methods in the same style as the other methods in this file

File: src/plasma/test/test.py [hunk 13]
```diff
@@ -71,6 +71,7 @@ def setUp(self):
       time.sleep(2.0)
     else:
       self.p = subprocess.Popen(command)
+      time.sleep(2.0)
```
Comment (robertnishihara) [inline src/plasma/test/test.py]: this will probably go away when you rebase

**Comment by justinyum** [comment]
Note - maybe put fsync() in persistence, don't use O_DIRECT flag for now.

**Comment by robertnishihara** [comment]
Closing for now, though we should still add this functionality at some point.

**Pull request closed without merging**